### PR TITLE
feat: [CI-5672]: cache intelligence steps

### PR DESCRIPTION
**Pull request by jamie-harness**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/38609)

### CODE REVIEW

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/commonconstants/CIExecutionConstants.java [hunk 1]
```diff
@@ -48,6 +48,14 @@ public class CIExecutionConstants {
   public static final String GIT_URL_SUFFIX = ".git";
   public static final String PATH_SEPARATOR = "/";
 
+  // COnstatns for implict cache steps
+  public static final String RESTORE_CACHE_STEP_ID = "restore-cache-harness";
+  public static final String SAVE_CACHE_STEP_ID = "save-cache-harness";
+  public static final String RESTORE_CACHE_STEP_NAME = "Restore Cache From Harness";
+  public static final String SAVE_CACHE_STEP_NAME = "Save Cache to Harness";
+  public static final String CACHE_ARCHIVE_TYPE_TAR = "tar";
```
Comment (rutvijmehta-harness) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/commonconstants/CIExecutionConstants.java:56]: Enum required?

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java [hunk 2]
```diff
@@ -71,6 +99,10 @@ public class CIStepGroupUtils {
   @Inject private CIExecutionConfigService ciExecutionConfigService;
   @Inject private CIExecutionServiceConfig ciExecutionServiceConfig;
   @Inject private VmInitializeTaskParamsBuilder vmInitializeTaskParamsBuilder;
+  @Inject private CIFeatureFlagService featureFlagService;
+
+  private static final String STRING_TRUE = "true";
+  private static final String STRING_FALSE = "false ";
```
Comment (rutvijmehta-harness) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java:105]: typo: "false"

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java [hunk 3]
```diff
@@ -89,15 +121,23 @@ public List<ExecutionWrapperConfig> createExecutionWrapperWithInitializeStep(Int
 
     List<ExecutionWrapperConfig> initializeExecutionSections = new ArrayList<>();
     boolean gitClone = RunTimeInputHandler.resolveGitClone(integrationStageConfig.getCloneCodebase());
-
+    CacheOptions cacheOptions = integrationStageConfig.getCacheOptions();
+    boolean saveCache =
+        cacheOptions != null && RunTimeInputHandler.resolveBooleanParameter(cacheOptions.getEnabled(), false);
+    CICacheIntelligenceConfig cacheIntelligenceConfig = ciExecutionServiceConfig.getCacheIntelligenceConfig();
```
Comment (rutvijmehta-harness) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java:127]: Need a not null check?

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java [hunk 4]
```diff
@@ -89,15 +121,23 @@ public List<ExecutionWrapperConfig> createExecutionWrapperWithInitializeStep(Int
...
+    boolean featureCacheEnabled = featureFlagService.isEnabled(CI_CACHE_INTELLIGENCE, accountId);
     if (gitClone) {
       initializeExecutionSections.add(
           getGitCloneStep(ciExecutionArgs, ciCodebase, accountId, IntegrationStageUtils.getK8OS(infrastructure)));
     }
-    for (ExecutionWrapperConfig executionWrapper : executionSections) {
-      initializeExecutionSections.add(executionWrapper);
+    if (featureCacheEnabled && saveCache) {
```
Comment (rutvijmehta-harness) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java:133]: We can merge L133 and L138 checks.

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java [hunk 5]
```diff
@@ -284,4 +324,118 @@ private ExecutionWrapperConfig getGitCloneStep(
...
+                              .envVariables(envVariables)
+                              .entrypoint(ParameterField.createValueField(entrypoint))
+                              .harnessManagedImage(true)
+                              .build();
+
+    PluginStepNode pluginStepNode =
+        PluginStepNode.builder()
+            .identifier(RESTORE_CACHE_STEP_ID)
+            .name(RESTORE_CACHE_STEP_NAME)
+            .timeout(ParameterField.createValueField(Timeout.builder().timeoutString("1h").build()))
```
Comment (rutvijmehta-harness) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java:356]: We should avoid magic string

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java [hunk 6]
```diff
@@ -284,4 +324,118 @@ private ExecutionWrapperConfig getGitCloneStep(
...
+  private ExecutionWrapperConfig getSaveCacheStep(
+      CacheOptions cacheOptions, String accountId, CICacheIntelligenceConfig cacheIntelligenceConfig) {
+    Map<String, JsonNode> settings = new HashMap<>();
+    Map<String, String> envVariables = new HashMap<>();
+    String uuid = generateUuid();
+    String saveCacheImage = ciExecutionConfigService.getPluginVersionForK8(SAVE_CACHE_GCS, accountId).getImage();
+    List<String> entrypoint = ciExecutionServiceConfig.getStepConfig().getCacheGCSConfig().getEntrypoint();
+
+    setCacheEnvVariables(envVariables, cacheIntelligenceConfig, cacheOptions, accountId);
+    envVariables.put(PLUGIN_OVERRIDE, String.valueOf(false));
```
Comment (rutvijmehta-harness) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java:379]: `STRING_FALSE`

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java [hunk 7]
```diff
@@ -284,4 +324,118 @@ private ExecutionWrapperConfig getGitCloneStep(
...
+                              .envVariables(envVariables)
+                              .entrypoint(ParameterField.createValueField(entrypoint))
+                              .harnessManagedImage(true)
+                              .build();
+
+    PluginStepNode pluginStepNode =
+        PluginStepNode.builder()
+            .identifier(SAVE_CACHE_STEP_ID)
+            .name(SAVE_CACHE_STEP_NAME)
+            .timeout(ParameterField.createValueField(Timeout.builder().timeoutString("1h").build()))
```
Comment (rutvijmehta-harness) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java:396]: Same as above.

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java [hunk 8]
```diff
@@ -284,4 +324,118 @@ private ExecutionWrapperConfig getGitCloneStep(
...
+      return ExecutionWrapperConfig.builder().uuid(uuid).step(jsonNode).build();
+    } catch (IOException e) {
+      throw new CIStageExecutionException("Failed to create restore cache step", e);
+    }
+  }
+
+  private void setCacheEnvVariables(Map<String, String> envVariables, CICacheIntelligenceConfig cacheIntelligenceConfig,
+      CacheOptions cacheOptions, String accountId) {
+    List<String> cacheDir = new ArrayList<>();
+    if (cacheOptions.getCachedPaths() != null) {
```
Comment (rutvijmehta-harness) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java:413]: We're already checking this condition in the caller

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java [hunk 9]
```diff
@@ -284,4 +324,118 @@ private ExecutionWrapperConfig getGitCloneStep(
...
+    envVariables.put(PLUGIN_BUCKET, cacheIntelligenceConfig.getBucket());
+    envVariables.put(PLUGIN_AUTO_DETECT_CACHE, STRING_TRUE);
+    envVariables.put(PLUGIN_AUTO_CACHE_ACCOUNT_ID, accountId);
+    if (cacheDir != null && cacheDir.size() > 0) {
+      envVariables.put(PLUGIN_MOUNT, listToStringSlice(cacheDir));
+    }
+    envVariables.put(PLUGIN_EXIT_CODE, STRING_TRUE);
+    envVariables.put(PLUGIN_ARCHIVE_FORMAT, CACHE_ARCHIVE_TYPE_TAR);
+    envVariables.put(PLUGIN_BACKEND, CACHE_GCS_BACKEND);
+    envVariables.put(PLUGIN_BACKEND_OPERATION_TIMEOUT, format("%ss", 10000));
```
Comment (rutvijmehta-harness) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java:426]: Define the timeout as a constant

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java [hunk 10]
```diff
@@ -284,4 +324,118 @@ private ExecutionWrapperConfig getGitCloneStep(
...
+    envVariables.put(PLUGIN_JSON_KEY, cacheIntelligenceConfig.getServiceKey());
+  }
+
+  private static String listToStringSlice(List<String> stringList) {
```
Comment (rutvijmehta-harness) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java:430]: `String.join(",", stringList);`

File: 956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java [hunk 11]
```diff
@@ -450,7 +450,8 @@ public enum FeatureName {
   CD_SERVICE_ENV_RECONCILIATION("Do reconciliation of service and env on pipeline/template save", HarnessTeam.CDC),
   CD_TRIGGER_CATALOG("Enables UI for Trigger catalog for Nexus ", HarnessTeam.CDC),
   PL_CONNECTOR_ENCRYPTION_PRIVILEGED_CALL("make the encryption/decryption call as pirvileged call", HarnessTeam.PL),
-  CDS_SHOW_CREATE_PR("Start showing CreatePR step on the plan creator if enabled", HarnessTeam.GITOPS);
+  CDS_SHOW_CREATE_PR("Start showing CreatePR step on the plan creator if enabled", HarnessTeam.GITOPS),
+  CI_CACHE_INTELLIGENCE("Feature flag fro cache intelligence feature", HarnessTeam.CI);
```
Comment (rutvijmehta-harness) [inline 956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java:454]: typo: for

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/commonconstants/CIExecutionConstants.java [hunk 12]
```diff
@@ -48,6 +48,14 @@ public class CIExecutionConstants {
   public static final String GIT_URL_SUFFIX = ".git";
   public static final String PATH_SEPARATOR = "/";
 
+  // COnstatns for implict cache steps
```
Comment (rutvijmehta-harness) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/commonconstants/CIExecutionConstants.java:51]: typo: Constants

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java (hunk 2 above)
Comment (jamie-harness) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java:105]: huh?

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/commonconstants/CIExecutionConstants.java (hunk 1 above)
Comment (jamie-harness) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/commonconstants/CIExecutionConstants.java:56]: no, this is jut static string to pass into plugin

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java (hunk 3 above)
Comment (jamie-harness) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java:127]: which one? Neither ciExecutionServiceConfig (injected) nor cacheIntelligenceConfig(can be empty) will be null

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java (hunk 4 above)
Comment (jamie-harness) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java:133]: no, the order matters

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java (hunk 8 above)
Comment (jamie-harness) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java:413]: No, I don't think it's being checked

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java (hunk 2 above)
Comment (rutvijmehta-harness) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java:105]: false string has an extra space

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java (hunk 3 above)
Comment (rutvijmehta-harness) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java:127]: Ack.

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java (hunk 10 above)
Comment (jamie-harness) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java:430]: this is copied over... I think I should change in the other place too

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java (hunk 2 above)
Comment (jamie-harness) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java]: ack

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java (hunk 5 above)
Comment (jamie-harness) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java]: done

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java (hunk 6 above)
Comment (jamie-harness) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java]: done

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java (hunk 7 above)
Comment (jamie-harness) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java]: done

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java (hunk 9 above)
Comment (jamie-harness) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java]: done

File: 956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java (hunk 11 above)
Comment (jamie-harness) [reply 956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java]: done

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/commonconstants/CIExecutionConstants.java (hunk 12 above)
Comment (jamie-harness) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/commonconstants/CIExecutionConstants.java]: done

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java (hunk 5 above)
Comment (jamie-harness) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/CIStepGroupUtils.java]: done

**Review by rutvijmehta-harness — COMMENTED** [review]
> Reviewed 5 of 14 files at r1, 1 of 3 files at r2, 2 of 4 files at r3, 9 of 9 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38609)* status: 7. unresolved discussions. 0 of 1 approvals obtained. Waiting on team(s): sto.

**Review by rutvijmehta-harness — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38609#-:-NECOHff7k8wZUEMx6Ra:bnfp4nl)* status: 7. unresolved discussions. Waiting on team(s): sto.

**Review by rutvijmehta-harness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38609)* status: Waiting on team(s): sto.

**Review by r0610205 — APPROVED** [review]

**Pull request merged**